### PR TITLE
470 Full experiment plan to replace execute request

### PR DESCRIPTION
**Pull request by rtuck99**
This fix addresses #470 

It also requires GDA jython script change (gerrit change request https://gerrit.diamond.ac.uk/c/gda/gda-mx/+/42628)

To test:

Confirm unit tests pass, and additional system test `test_load_centre_collect_full_plan` work
This code change should support single and multiple rotations, and will be enabled if the property
`gda.mx.hyperion.load_centre_collect.enabled` is set to `true` in domain.properties

### CODE REVIEW

File: tests/unit_tests/hyperion/experiment_plans/test_load_centre_collect_full_plan.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+def find_a_pin(pin_tip_detection):
+    def set_good_position():
+        set_mock_value(pin_tip_detection.triggered_tip, (100, 110))
+        return NullStatus()
+
+    return set_good_position
+
+
+def dont_find_pin(pin_tip_detection):
```
Comment (DominicOram) [inline tests/unit_tests/hyperion/experiment_plans/test_load_centre_collect_full_plan.py:41]: Should: I don't think this is used? Can we remove it?

File: tests/system_tests/hyperion/external_interaction/test_load_centre_collect_full_plan.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+
+def compare_comment(
+    fetch_datacollection_attribute, data_collection_id, expected_comment
+):
+    actual_comment = fetch_datacollection_attribute(
+        data_collection_id, DATA_COLLECTION_COLUMN_MAP["comments"]
+    )
+    match = re.search(" Zocalo processing took", actual_comment)
+    truncated_comment = actual_comment[: match.start()] if match else actual_comment
+    assert truncated_comment == expected_comment
```
Comment (DominicOram) [inline tests/system_tests/hyperion/external_interaction/test_load_centre_collect_full_plan.py:305]: Should: Are these exactly the same as those in `test_ispyb_dev_connection.py`? Can we move them to somewhere common?

File: tests/system_tests/hyperion/external_interaction/test_load_centre_collect_full_plan.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    "slitGapVertical": 0.234,
+    "xtalSnapshotFullPath1": "/tmp/snapshot6.png",
+    "xtalSnapshotFullPath2": "/tmp/snapshot7.png",
+    "xtalSnapshotFullPath3": "/tmp/snapshot8.png",
+    "xtalSnapshotFullPath4": "/tmp/snapshot9.png",
+}
+
+
+@pytest.mark.s03
+def test_execute_load_centre_collect_full_plan(
```
Comment (DominicOram) [inline tests/system_tests/hyperion/external_interaction/test_load_centre_collect_full_plan.py:184]: Must: Running this I get an error:
```
FAILED tests/system_tests/hyperion/external_interaction/test_load_centre_collect_full_plan.py::test_execute_load_centre_collect_full_plan - AssertionError: 
  expected beamSizeAtSampleX 0.1 == 0.02
  
assert '\nexpected beamSizeAtSampleX 0.1 == 0.02\n' == '\n'
  
  Strings contain only whitespace, escaping them using repr()
  - '\n'
  + '\nexpected beamSizeAtSampleX 0.1 == 0.02\n'
```

Comment (DominicOram) [reply tests/system_tests/hyperion/external_interaction/test_load_centre_collect_full_plan.py:184]: Otherwise, love this test though, thank you!

**Review by DominicOram — CHANGES_REQUESTED** [review]
This is great, thank you! I think the only major issue is the `WarningException` thing otherwise just cleaning up. I think the failing test and some of the linting can be fixed by https://github.com/DiamondLightSource/mx-bluesky/pull/506/files.

File: src/mx_bluesky/hyperion/experiment_plans/load_centre_collect_full_plan.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+):
+    """Attempt a complete data collection experiment, consisting of the following:
+    * Load the sample if necessary
+    * Move to the specified goniometer start angles
+    * Perform optical centring, then X-ray centring
+    * If X-ray centring finds a diffracting centre then move to that centre and
+    * do a collection with the specified parameters.
+    """
+    if not oav_params:
+        oav_params = OAVParameters(context="xrayCentring")
```
[the remark this answers is not in the document]
Comment (DominicOram) [reply src/mx_bluesky/hyperion/experiment_plans/load_centre_collect_full_plan.py:43]: https://github.com/DiamondLightSource/mx-bluesky/issues/314

File: src/mx_bluesky/hyperion/parameters/components.py [hunk 5]
```diff
@@ -147,6 +147,15 @@ class WithOptionalEnergyChange(BaseModel):
...
     visit: str = Field(min_length=1)
+    zocalo_environment: str = Field(default=CONST.ZOCALO_ENV)
+    beamline: str = Field(default=CONST.I03.BEAMLINE, pattern=r"BL\d{2}[BIJS]")
+    det_dist_to_beam_converter_path: str = Field(
+        default=CONST.PARAM.DETECTOR.BEAM_XY_LUT_PATH
+    )
+    insertion_prefix: str = Field(
+        default=CONST.I03.INSERTION_PREFIX, pattern=r"SR\d{2}[BIJS]"
+    )
+    detector_distance_mm: float | None = Field(default=None, gt=0)
```
[the remark this answers is not in the document]
Comment (DominicOram) [reply src/mx_bluesky/hyperion/parameters/components.py]: https://github.com/DiamondLightSource/mx-bluesky/issues/540

File: src/mx_bluesky/hyperion/experiment_plans/flyscan_xray_centre_plan.py [hunk 6]
```diff
@@ -190,47 +196,51 @@ def run_gridscan_and_move(
...
-        LOGGER.info("Recentring smargon co-ordinate system to this point.")
-        yield from bps.mv(
-            fgs_composite.sample_motors.stub_offsets, StubPosition.CURRENT_AS_CENTER
-        )
-
-    # Turn off dev/shm streaming to avoid filling disk, see https://github.com/DiamondLightSource/hyperion/issues/1395
-    LOGGER.info("Turning off Eiger dev/shm streaming")
-    yield from bps.abs_set(fgs_composite.eiger.odin.fan.dev_shm_enable, 0)
+            else:
+                xray_centre = initial_xyz
```
[the remark this answers is not in the document]
Comment (DominicOram) [reply src/mx_bluesky/hyperion/experiment_plans/flyscan_xray_centre_plan.py:214]: https://github.com/DiamondLightSource/mx-bluesky/issues/541

File: src/mx_bluesky/hyperion/experiment_plans/flyscan_xray_centre_plan.py [hunk 7]
```diff
@@ -190,47 +196,51 @@ def run_gridscan_and_move(
...
+        with TRACER.start_span("move_to_result"):
+            x, y, z = xray_centre
+            yield from move_x_y_z(fgs_composite.sample_motors, x, y, z, wait=True)
+
+        if parameters.FGS_params.set_stub_offsets:
+            LOGGER.info("Recentring smargon co-ordinate system to this point.")
+            yield from bps.mv(
+                fgs_composite.sample_motors.stub_offsets, StubPosition.CURRENT_AS_CENTER
+            )
+    finally:
```
[the remark this answers is not in the document]
Comment (rtuck99) [reply src/mx_bluesky/hyperion/experiment_plans/flyscan_xray_centre_plan.py:236]: I will look at doing this for the multipin changes as I will need to move that anyway

File: tests/system_tests/hyperion/external_interaction/conftest.py [hunk 8]
```diff
@@ -167,3 +209,158 @@ def dummy_ispyb_3d(dummy_params) -> StoreInIspyb:
...
+        undulator=undulator_for_system_test,
+        aperture_scatterguard=aperture_scatterguard,
+        synchrotron=synchrotron,
+        s4_slit_gaps=s4_slit_gaps,
+        zebra=zebra,
+        robot=robot,
+        oav=oav_for_system_test,
+        sample_shutter=sample_shutter,
+        xbpm_feedback=xbpm_feedback,
+    )
```
[the remark this answers is not in the document]
Comment (rtuck99) [reply tests/system_tests/hyperion/external_interaction/conftest.py:340]: The two fixtures use different sets of devices - `oav_for_system_test` and `undulator_for_system_test`

File: tests/unit_tests/hyperion/experiment_plans/test_load_centre_collect_full_plan.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+from ....conftest import pin_tip_edge_data, raw_params_from_file
+
+
+def find_a_pin(pin_tip_detection):
+    def set_good_position():
+        set_mock_value(pin_tip_detection.triggered_tip, (100, 110))
+        return NullStatus()
+
+    return set_good_position
```
[the remark this answers is not in the document]
Comment (rtuck99) [reply tests/unit_tests/hyperion/experiment_plans/test_load_centre_collect_full_plan.py:38]: `dont_find_pin` isn't used anyway as in the end I rewrote the test using the run engine simulator so I will remove it

File: tests/system_tests/hyperion/external_interaction/test_load_centre_collect_full_plan.py (hunk 2 above)
Comment (rtuck99) [reply tests/system_tests/hyperion/external_interaction/test_load_centre_collect_full_plan.py:305]: Yes, I have moved them

File: tests/unit_tests/hyperion/experiment_plans/test_load_centre_collect_full_plan.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (DominicOram) [reply tests/unit_tests/hyperion/experiment_plans/test_load_centre_collect_full_plan.py:38]: I still think the use of `partial` and `AsyncStatus` is cleaner than the nested functions and `NullStatus` but not worth holding things up on

**Pull request merged**